### ISSUE TITLE
LLama 2 compatibility 

### ISSUE THREAD

**Issue by chris-aeviator** [Ber666/llm-reasoners#18]
Given the restricted access to LLama 1, is it possible to use llm-reasoner examples with LLama 2 out of the box?

**Comment by Ber666** [comment]
Hi, thanks for the suggestion! We are working on supporting Llama-2, and hopefully, it'll be ready within one week!

**Comment by hilarious-viking** [comment]
Executing now [gsm8k](https://paperswithcode.com/sota/arithmetic-reasoning-on-gsm8k) using [llama.cpp](https://github.com/ggerganov/llama.cpp): [wizardlm-13b-v1.2.ggmlv3.q8_0](https://huggingface.co/TheBloke/WizardLM-13B-V1.2-GGML) (fine tuned Llama 2)
```
Case #1: correct=True, output='18', answer='18' ; accuracy=1.000 (1/1)
...
Case #100: correct=False, output='40', answer='58' ; accuracy=0.530 (53/100)
Case #101: correct=False, output='85', answer='175' ; accuracy=0.525 (53/101)
Case #102: correct=False, output='24', answer='6' ; accuracy=0.520 (53/102)
Case #103: correct=False, output='15', answer='26' ; accuracy=0.515 (53/103)
Case #104: correct=True, output='140', answer='140' ; accuracy=0.519 (54/104)
Case #105: correct=False, output='250', answer='500' ; accuracy=0.514 (54/105)
Case #106: correct=True, output='20', answer='20' ; accuracy=0.519 (55/106)
Case #107: correct=False, output='360', answer='72' ; accuracy=0.514 (55/107)
Case #108: correct=False, output='2', answer='3' ; accuracy=0.509 (55/108)
Case #109: correct=False, output='111', answer='50' ; accuracy=0.505 (55/109)
Case #110: correct=True, output='28', answer='28' ; accuracy=0.509 (56/110)
```
that places results somewhere around LLaMA 65B/Text-davinci-002-175B, impressive considering that [Llama 2 70B GSM8K (8-shot) gets 56.8](https://arxiv.org/pdf/2307.09288.pdf), interesting to see results of Llama 2 70B with this method...

Few questions:
- why temperature 0.8 (in some cases 1.0) is used ? for reproducible results is it better to use 0.0, no? 
- any plans to provide [langchain](https://github.com/langchain-ai/langchain) integration, would be nice to see agent of that type there

Thank you !

**Comment by Ber666** [comment]
> Executing now [gsm8k](https://paperswithcode.com/sota/arithmetic-reasoning-on-gsm8k) using [llama.cpp](https://github.com/ggerganov/llama.cpp): [wizardlm-13b-v1.2.ggmlv3.q8_0](https://huggingface.co/TheBloke/WizardLM-13B-V1.2-GGML) (fine tuned Llama 2: [WizardLM](https://github.com/nlpxucan/WizardLM))
> 
> ```
> Case #1: correct=True, output='18', answer='18' ; accuracy=1.000 (1/1)
> ...
> Case #115: correct=False, output='50', answer='6' ; accuracy=0.513 (59/115)
> Case #116: correct=False, output='80', answer='90' ; accuracy=0.509 (59/116)
> Case #117: correct=False, output='221.2', answer='42' ; accuracy=0.504 (59/117)
> Case #118: correct=True, output='360', answer='360' ; accuracy=0.508 (60/118)
> Case #119: correct=False, output='8', answer='4' ; accuracy=0.504 (60/119)
> ```
> 
> that places results somewhere around LLaMA 65B/Text-davinci-002-175B, impressive considering that [Llama 2 70B GSM8K (8-shot) gets 56.8](https://arxiv.org/pdf/2307.09288.pdf), interesting to see results of Llama 2 70B with this method...
> 
> Few questions:
> 
> * why temperature 0.8 is used ? for reproducible results is it better to use 0.0, no?
> * to solve Math questions RAP was modified from state/action/reward perspectives, if that is the case, is it possible to generalise RAP ? One RAP to rule the world, or it's a theme for next paper ?)
> * I see that evolution of gsm8k is pretty slow, I suppose it's because of RAP-Aggregate ?
> * any plans to provide [langchain](https://github.com/langchain-ai/langchain) integration, would be nice to see agent of that type there
> 
> Thank you !

Hi, thank you for sharing the results!

- Our method needs to randomly sample multiple different subquestions/subanswers, so that we can search for the optimal reasoning path. If using 0.0 temperature, all the generated contents will be the same and we won't have candidates to choose from.
- Great question! I think there is a trade-off here: The most general formulation of a world model is to simply append the new action to a state: $s_t = [s_{t-1}; a_t]$, and we can still use advanced search algorithms and achieve non-trivial performance, as shown by the other examples ([1](https://github.com/Ber666/llm-reasoners/tree/main/examples/guided_gsm8k), [2](https://github.com/Ber666/llm-reasoners/tree/main/examples/tot_game24)). However, to fully take advantage of RAP, it's preferred to formulate problems in a specific way, e.g. for [blocksworld](https://github.com/Ber666/llm-reasoners/tree/main/examples/rap_blocksworld), define the state as the block condition will be much more beneficial. 
- It's because the (reward calculation](https://github.com/Ber666/llm-reasoners/blob/43e241f8917310d82d489dec98d9c9823dc889ee/examples/rap_gsm8k/world_model.py#L68C34-L68C34) involves sampling multiple answers and doing a majority vote. We find it to be a pretty reliable reward, but the price is its cost. This shouldn't be relevant to RAP-Aggregate. We haven't included the script to do aggregation, but it's only a simple traversal of the saved tree and won't take much time.
- Thanks for your advice, but for now I am not sure about how we can integrate langchain and what LLM Reasoners can benefit from it. Any ideas will be appreciated.

**Comment by hilarious-viking** [comment]
Thank you @Ber666 for answers, after playing a little with RAP, it's totally make sense.

May I suggest for Figure 7, RAP(1) is getting score 40 with 1 iteration, essentially it's means Few-shots with k=4 + CoT (? 1 iteration means you asked 1 question, get response, than you ask for final answer). So, only from that point (score 40) you are getting advantage from RAP. And in this regards you need to show on chart Few-shots with k=0,1,2,3,4, to see relative RAP gains, otherwise it's misleading, suggesting that RAP can achieve great result with 1 iteration, where all gains comes from Few-shots.

But regardless of that suggest, I would like to acknowledge that results are impressive, with very small subset of GSM8K and [StableBeluga2-70B](https://huggingface.co/stabilityai/StableBeluga2) + CoT using langchain + tools (WolframAlpha, python, wikipedia, search etc.) I got around 50% ~= similar to yesterday posted [wizardlm-13b-v1.2.ggmlv3.q8_0](https://huggingface.co/TheBloke/WizardLM-13B-V1.2-GGML) + RAP

Example output of StableBeluga2-70B + langchain + tools: 
```
Question: Every day, Wendi feeds each of her chickens three cups of mixed chicken feed, containing seeds, mealworms and vegetables to help keep them healthy.  She gives the chickens their feed in three separate meals. In the morning, she gives her flock of chickens 15 cups of feed.  In the afternoon, she gives her chickens another 25 cups of feed.  How many cups of feed does she need to give her chickens in the final meal of the day if the size of Wendi's flock is 20 chickens?
Thought: I should calculate how much food each chicken gets per meal and then find out how much food they get in total for the first two meals. Then, I can subtract that from the total amount of feed given to all chickens (60 cups) to find out how much is left for the final meal.
Action: WolframAlpha
Action Input: 15 + 25 =
Observation: Assumption: 15 + 25 
Answer: 40
Thought: Now I know that Wendi gives her chickens a total of 40 cups of feed in the first two meals.
Action: WolframAlpha
Action Input: 3 * 20 =
Observation: Assumption: 3×20 
Answer: 60
Thought: Now I know that Wendi gives her chickens a total of 60 cups of feed per day.
Action: WolframAlpha
Action Input: 40 / 20 =
Observation: Assumption: 40/20 
Answer: 2
Thought: Now I know that each chicken gets 2 cups of food per meal.
Action: WolframAlpha
Action Input: 60 - 40 =
Observation: Assumption: 60 - 40 
Answer: 20
Thought:
Generated:
 Now I know that Wendi needs to give her chickens a total of 20 cups of feed in the final meal.
Final Answer: Wendi needs to give her chickens 20 cups of feed in the final meal of the day.
```
![image](https://github.com/Ber666/llm-reasoners/assets/27341699/4c28429e-cbc0-4bbb-90dc-913ab1e2672a)

I'm thinking about mixing langchain tools with RAP, so instead of Few-shots, provide tools available. Llama 2 is much better with logic, I suspect that Few-shots could be less important for it than for Llama 1, testing now RAP with k=0 and 1
Your project can benefit from it as it's very high profile project, as llama.cpp is, if your method increase reasoning capabilities on small models (7B/13B), it will be highly used as it can run on commodity hardware (24G vram), and having more users, means more stars for your project, paper citation, projects will start to optimise perf for your method (indeed eval in llama.cpp is pretty slow, I suspect it's not done on CUDA kernel, but CPU only)

**Comment by Ber666** [comment]
Hi @hilarious-viking, thanks for your questions. I would like to clarify some points:

> 1 iteration means you asked 1 question, get response, than you ask for final answer
No. One iteration will lead to a chain without any branches (instead of a tree), but it still involves multiple steps of reasoning.
> only from that point (score 40) you are getting advantage from RAP
Not exactly. It's fair to say that only from that point we are getting advantage from Monte-carlo Tree Search, but the RAP framework also include a world model formulation for numerical reasoning, i.e. asking subquestions and answer them in turn. It'll be beneficial even if we only use 1 iteration.

May I ask which argument are you talking about when you say `k=...`?

**Comment by chris-aeviator** [comment]
I managed to get the llama 2 model "running" via transformers, needed to replace attention_masks or set them to None, and some changes in the logits processing, but this logits effort broke the successful prediction. 

I'm completely impressed by LLama 1 predictions, basically the rap_gsm8 example  allows to adapt to all kinds of reasoning scenarios (non-math) with  high quality reasoning on a 7b model with little code & prompt changes.

**Comment by hilarious-viking** [comment]
Hello @Ber666, thanks again for clarifications,

> No. One iteration will lead to a chain without any branches (instead of a tree), but it still involves multiple steps of reasoning.
Got it, indeed 1 iter is one graph evaluation, I suppose the one dumped in .pkl is the one taken as solution but in reality graph is bigger, is it possible to easily dump full graph ?

> May I ask which argument are you talking about when you say k=...?
in Few-shots k is number of examples you are showing in your prompt before asking your question, in gsm8k you are providing 4 examples (interactive_examples.json & useful_examples.json), 5th is gsm8k question itself, that why 40 base score is Few-shots k=4 w/ RAP(1), that why would be interesting to see results with k=0,1,2,3 too, to isolate just RAP effectiveness from Few-shots (it's known that Few-shots provide good results to begin with)

>  basically the rap_gsm8 example allows to adapt to all kinds of reasoning scenarios (non-math) with high quality reasoning
@chris-aeviator doing the same, but with llama 2 13B, llama 2 is surprisingly good at following instructions

**Comment by hilarious-viking** [comment]
@Ber666 executed RAP (1) w/o Few-shots, score stabilises around: 0.23
vs. 0.47 RAP (10) w/ Few-shots k=4
Same mode: [wizardlm-13b-v1.2.ggmlv3.q8_0](https://huggingface.co/TheBloke/WizardLM-13B-V1.2-GGML)
params: temperature: 0.8 , top_k: 40 , top_p: 0.95

Without Few-shots, responses could be without asked "The answer is", below Case #184 could not be parsed because answer was: ["The answer is that they will win the race by two seconds."](https://www.llm-reasoners.net/visualizer/4225db9e-4789-4325-966f-185769e50af5?accessKey=4ffa8f73), which is actually correct, there just few empty results, I have checked them, and most not correct anyway, Case #184 is exception, so final score could be ~0.25, but it's far from what you have in paper 0.40
Executing now RAP (1) w/ Few-shots k=4...

RAP (1) w/o Few-shots:
```
Case #180: correct=True, output='130', answer='130' ; accuracy=0.228 (41/180)
Case #181: correct=False, output='60', answer='20' ; accuracy=0.227 (41/181)
Case #182: correct=True, output='160', answer='160' ; accuracy=0.231 (42/182)
Case #183: correct=False, output='32', answer='23' ; accuracy=0.230 (42/183)
Case #184: correct=False, output='', answer='2' ; accuracy=0.228 (42/184)
Case #185: correct=False, output=None, answer='25' ; accuracy=0.227 (42/185)
Case #186: correct=True, output='30', answer='30' ; accuracy=0.231 (43/186)
Case #187: correct=False, output='6', answer='5' ; accuracy=0.230 (43/187)
Case #188: correct=True, output='106', answer='106' ; accuracy=0.234 (44/188)
Case #189: correct=False, output='20', answer='50' ; accuracy=0.233 (44/189)
Case #190: correct=False, output='29.50', answer='34' ; accuracy=0.232 (44/190)
Case #191: correct=False, output='36', answer='360' ; accuracy=0.230 (44/191)
Case #192: correct=False, output='35', answer='5' ; accuracy=0.229 (44/192)
Case #193: correct=False, output=None, answer='91' ; accuracy=0.228 (44/193)
Case #194: correct=False, output='18', answer='24' ; accuracy=0.227 (44/194)
Case #195: correct=False, output='0', answer='10' ; accuracy=0.226 (44/195)
Case #196: correct=True, output='12.00', answer='12' ; accuracy=0.230 (45/196)
Case #197: correct=False, output='3000', answer='120' ; accuracy=0.228 (45/197)
```

interactive_0_example.json
```
{
    "input": "Given a question, please decompose it into sub-questions. For each sub-question, please answer it in a complete sentence, ending with \"The answer is\". When the original question is answerable, please start the subquestion with \"Now we can answer the question: \".\n\n",
    "question_prefix": "Question 1: ",
    "subquestion_prefix": "Question 1.{}:",
    "overall_question_prefix": "Now we can answer the question:",
    "answer_prefix": "Answer 1.{}:",
    "index": 1
}
```

useful_0_example.json
```
{
  "input": "Given a question and some sub-questions, determine whether the last sub-question is useful to answer the question. Output 'Yes' or 'No', and a reason.\n\n",
  "question_prefix": "Question 1: ",
  "subquestion_prefix": "Question 1.{}:",
  "new_subquestion_prefix": "New question 1.{}:",
  "useful_prefix": "Is the new question useful?"
}
```

**Comment by AegeanYan** [comment]
@hilarious-viking Thanks for your kindly help. We have now updated for llama 2's integration the main branch. Hope it will be helpful for you. If anyone have question in experiment with llama 2, please directly contact with me.

**Comment by AegeanYan** [comment]
> so without Few-shots RAP would be likely lower than CoT + CS for gsm8k

I'm reproducing the work in the paper recently and I have to say that RAP's comparison is reasonable. Because the result in the paper for CoT and SC is also on 4-shot.
